statistics/handle/lru_cache.go

```diff
@@ -0,0 +1,336 @@
+	l.Unlock()
+	if calculated {
+		metrics.StatsCacheLRUMemUsage.WithLabelValues(typTrack).Set(float64(trackingCost))
+		metrics.StatsCacheLRUMemUsage.WithLabelValues(typTotal).Set(float64(totalCost))
+	}
+}
+
+func (l *internalLRUCache) calculateTableCost(key int64) bool {
+	element, exists := l.elements[key]
+	if !exists {
+		return false
+	}
+	item := element.Value.(*cacheItem)
+	l.put(item.key, item.value, item.value.MemoryUsage(), true)
+	return true
+}
+
+// Copy returns a replication of LRU
+func (l *internalLRUCache) Copy() statsCacheInner {
+	var newCache *internalLRUCache
+	l.RLock()
+	newCache = newInternalLRUCache(l.capacity)
+	node := l.cache.Back()
+	for node != nil {
+		key := node.Value.(*cacheItem).key
+		value := node.Value.(*cacheItem).value
+		tblMemUsage := node.Value.(*cacheItem).tblMemUsage
+		newCache.put(key, value, tblMemUsage, false)
+		node = node.Prev()
+	}
+	l.RUnlock()
+	metrics.StatsCacheLRUCounter.WithLabelValues(typCopy).Inc()
+	return newCache
+}
+
+// internalLRUCache will evict a table's column' structure in order to keep tracking cost under capacity
+// If the elements has no structure can be evicted, it will be removed from list.
+func (l *internalLRUCache) evictIfNeeded() {
+	curr := l.cache.Back()
+	evicted := false
+	for !l.underCapacity() {
+		evicted = true
+		item := curr.Value.(*cacheItem)
+		tbl := item.value
+		oldMemUsage := item.tblMemUsage
+		prev := curr.Prev()
+		for _, col := range tbl.Columns {
+			if col.IsEvicted() {
+				continue
+			}
+			col.DropEvicted()
+			newMemUsage := tbl.MemoryUsage()
+			item.tblMemUsage = newMemUsage
+			l.calculateCost(newMemUsage, oldMemUsage)
+			if l.underCapacity() {
+				break
+			}
+		}
+		newMemUsage := tbl.MemoryUsage()
+		if newMemUsage.TotalColTrackingMemUsage() < 1 {
+			l.maintainList(curr, nil, newMemUsage.TotalColTrackingMemUsage(), oldMemUsage.TotalColTrackingMemUsage())
+		}
+		curr = prev
+	}
+	if evicted {
+		metrics.StatsCacheLRUCounter.WithLabelValues(typEvict).Inc()
+	}
+}
+
+func (l *internalLRUCache) calculateCost(newUsage, oldUsage *statistics.TableMemoryUsage) {
+	l.totalCost += newUsage.TotalMemUsage - oldUsage.TotalMemUsage
+	l.trackingCost += newUsage.TotalColTrackingMemUsage() - oldUsage.TotalColTrackingMemUsage()
+}
+
+// maintainList maintains elements in list cache
+// For oldTotalColTrackingMemUsage>0 && newTotalColTrackingMemUsage>0, it means the element is updated.
+// For oldTotalColTrackingMemUsage>0 && newTotalColTrackingMemUsage=0, it means the element is removed.
+// For oldTotalColTrackingMemUsage=0 && newTotalColTrackingMemUsage>0, it means the new element is inserted
+// For oldTotalColTrackingMemUsage=0 && newTotalColTrackingMemUsage=0, we do nothing.
+func (l *internalLRUCache) maintainList(element *list.Element, item *cacheItem, newTotalColTrackingMemUsage, oldTotalColTrackingMemUsage int64) *list.Element {
+	if oldTotalColTrackingMemUsage > 0 {
+		if newTotalColTrackingMemUsage > 0 {
+			l.cache.MoveToFront(element)
```

Comment:
I think we don't need to MoveToFront for update. It could be the table stats is updated by auto-analyze, it means nothing to LRU. 
And I think oldTotalColTrackingMemUsage=0 doesn't mean the element does not exist, it could be the CMSketch of this table are all evicted. 
I think we don't need to maintainList with memUsage, just do PushFront for new inserted table, do Remove for deleted table, and do MoveToFront for table GetByQuery.

Reply:
Just keep LRU list consistent with map, the only usage of LRU list is to provide the order for evict.